lib/puppet/type/iis_site.rb

```diff
@@ -314,6 +315,20 @@ def insync?(is)
     end
   end
 
+  newproperty(:authenticationinfo, :parent => PuppetX::PuppetLabs::IIS::Property::Hash) do
+    desc 'Enable and disable authentication schemas'
+    valid_schemas = ['anonymous', 'basic', 'clientCertificateMapping',
+                      'digest', 'iisClientCertificateMapping', 'windows']
+    validate do |value|
+      unless value.is_a?(Hash)
+          fail("All bindings must be a hash")
+      end
+      unless (valid_schemas - value.keys).empty?
```

Comment:
How do we check that any of the strings defined in valid_schemas are included in value.keys. We need to ensure at least one is there, or all. We need to ensure that anything else causes a validation failure

Reply:
```unless valid_schemas.include?(value)``` ?

Reply:
I don't think that will work because `value` is the hash, not each key in the hash.

Reply:
Other attempts:

```
[1] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> value.keys
=> ["basic", "anonymous"]
[2] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas.include?(value.keys)
=> false
[3] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas.include?('basic')
=> true
[4] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> (valid_schemas - value.keys).empty?
=> false
[5] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas - value.keys
=> ["clientCertificateMapping", "digest", "iisClientCertificateMapping", "windows"]
[6] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas - ['wakka']
=> ["anonymous", "basic", "clientCertificateMapping", "digest", "iisClientCertificateMapping", "windows"]
[7] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> (valid_schemas - ['wakka']).empty?
=> false
[8] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas.include?('wakka')
=> false
[9] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas.include?(['wakka','basic'])
=> false
[10] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas.include?(['basic'])
=> false
[11] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas
=> ["anonymous", "basic", "clientCertificateMapping", "digest", "iisClientCertificateMapping", "windows"]
[12] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> value.keys
=> ["basic", "anonymous"]
[13] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas & value.keys
=> ["anonymous", "basic"]
[14] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> (valid_schemas & value.keys).empty?
=> false
[15] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> (valid_schemas & ['wakka','basic').empty?
SyntaxError: unexpected ')', expecting ']'
(valid_schemas & ['wakka','basic').empty?
                                  ^
[15] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> (valid_schemas & ['wakka','basic']).empty?
=> false
[16] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> (valid_schemas & ['wakka']).empty?
=> true
[17] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas & ['wakka']
=> []
[18] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas & ['wakka','basic']
=> ["basic"]
[19] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas & value.keys
=> ["anonymous", "basic"]
[20] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> value.keys
=> ["basic", "anonymous"]
[21] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas.any? { |s| value.keys.include?(s) }
=> true
[22] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas.any? { |s| ['wakka'].include?(s) }
=> false
[23] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> valid_schemas.any? { |s| ['wakka','basic'].include?(s) }
=> true
[24] pry(#<Puppet::Type::Iis_site::Authenticationinfo>)> value
=> {"basic"=>true, "anonymous"=>false}
```

Reply:
I'm not coming up with an elegant way (one that avoids a loop) offhand.

Maybe something like

```ruby
unless (value.keys.map { |key| valid_schemas.include?(key) }).uniq == [true]
```

That's pretty ugly and confusing though.

Reply:
Ruby has some operators for doing set math - the `-` in this case is doing a difference between the two lists of things. That can work, but this sort of check is usually performed by comparing the set intersection of the specified and allowed items (using the `&` operator), then comparing it back to the specified like this:

```
[1] pry(main)> allowed = ['ethan', 'james', 'craig', 'michael']
=> ["ethan", "james", "craig", "michael"]

[2] pry(main)> specified = ['ethan', 'bill']
=> ["ethan", "bill"]

[3] pry(main)> specified & allowed
=> ["ethan"]

[4] pry(main)> (specified & allowed) == specified
=> false

[5] pry(main)> specified - allowed
=> ["bill"]
```

So when line 4 is `true`, the `specified` list only contains items from the `allowed` list, but in the above case we've specified an extra item that's not allowed, at which point you can do the set difference with `-` to create an error message with the disallowed names.

Also note that the sets are unordered, so the following still works just fine even though the items in `specified` are not the same order as `allowed`:

```
[6] pry(main)> specified = ['craig', 'james']
=> ["craig", "james"]
[7] pry(main)> specified & allowed
=> ["craig", "james"]
[8] pry(main)> (specified & allowed) == specified
=> true
```

Reply:
Our group provider has a similar check for `insync?` - https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/user/windows_adsi.rb#L38-L49